esp/services/ws_topology/ws_topologyService.cpp

```diff
@@ -539,64 +540,118 @@ bool CWsTopologyEx::findTimestampAndLT(const char *logname, IFile* rFile, ReadLo
+    if (rFile->size() < (readLogReq.logDatePos + LOG_DATE_TIME_LENGTH))
         return false;//no timestamp in this log
+    if ((readLogReq.logfields == 0) && !readLogTime(pTr, readLogReq.logDatePos, LOG_DATE_TIME_LENGTH, latestLogTime))
+        return false;//Only legacy log file (readLogReq.logfields = 0) may not have timestamp.
 
     if ((readLogReq.filterType != GLOLastNHours) && (readLogReq.filterType != GLOTimeRange))
         return true;//Not interested in the last timestamp this time
 
-    //Search the last chuck to find out the latest timestamp
-    if (readSize < fileSize)
+    return readLastLogTime(logname, rIO, fileSize, readSize, readLogReq, latestLogTime);
+}
+
+void CWsTopologyEx::readLogMessageFields(char* contentStart, size32_t bytesRemaining, ReadLog& readLogReq)
+{
+    //Find out the first log line
+    char* finger = contentStart;
+    while (bytesRemaining > 0)
     {
-        bytesRead = rIO->read(fileSize - readSize, readSize, dataBuffer.clear().reserve(readSize));
-        if (bytesRead != readSize)
-            throw MakeStringException(ECLWATCH_CANNOT_READ_FILE, "Failed to read file %s.", logname);
-        pTr = (char*) dataBuffer.str();
+        if (isLineTerminator(finger, bytesRemaining, readLogReq.ltBytes))
+            break;
+
+        finger++;
+        bytesRemaining--;
     }
 
-    //skip the first line which may not contain the entire log line
-    StringBuffer logLine;
-    size32_t bytesRemaining = bytesRead;
-    bool hasAnotherLine = false;
-    pTr = readALogLine(pTr, bytesRemaining, readLogReq.ltBytes, logLine, hasAnotherLine);
-    if (!hasAnotherLine || (bytesRemaining < 28))
-        return true;
+    if (finger > contentStart)
+    {//Try to read LogMessageFields from the first log line
+        StringBuffer logLine;
+        logLine.append(finger - contentStart, contentStart);
+        readLogReq.logfields = getMessageFieldsFromHeader(logLine);
+        if (readLogReq.logfields > 0)
+        {
+            readLogReq.logDatePos = getPositionOfField(readLogReq.logfields, MSGFIELD_date);
+            readLogReq.logTimePos = getPositionOfField(readLogReq.logfields, MSGFIELD_milliTime);
+        }
+    }
+}
 
-    //Find out the last timestamp
-    while (hasAnotherLine && bytesRemaining > 27)
+bool CWsTopologyEx::readLastLogTime(const char *logName, IFileIO* rIO, size32_t fileSize, size32_t readSize, ReadLog& readLogReq, CDateTime& latestLogTime)
+{
+    size32_t readFrom = 0;
+    if (readSize < fileSize)
+        readFrom = fileSize - readSize; //Search from the last chuck
+
+    StringBuffer contentBuffer, previousPartialLine;
+    while (1)
     {
-        CDateTime dt;
-        pTr = readALogLine(pTr, bytesRemaining, readLogReq.ltBytes, logLine, hasAnotherLine);
-        if ((logLine.length() > 27) && readLogTime((char*) logLine.str(), 9, 19, dt))
-            latestLogTime = dt;
+        size32_t bytesRead = rIO->read(readFrom, readSize, contentBuffer.clear().reserve(readSize));
```

Comment:
This looks like it will read the whole of the rest of the file, just to read the last log line?  If so that seems inefficient - it could be better to read near the end of the file, and work backwards (assuming there are new newlines within log lines - although I'm not sure this code would cope with newlines in log lines anyway).